### ISSUE TITLE
weird parsing thing

### ISSUE THREAD

**Issue by lolbinarycat** [fish-shell/fish-shell#10754]
fish, version 3.7.1

type the following into a prompt and hit enter
```
begin
echo a & |
```

expected behavior: newline

actual behavior: `fish: Missing end to balance this begin` error

**Comment by faho** [comment]
`&` is the background operator, it is also a statement terminator. `echo a; |` triggers the same error.

What this is pushing you towards is

```fish
begin
   echo a &
end | foo
```

where you were expecting it to push you to

```fish
begin
    echo a &
    foo | bar
```

or for the `&` to work differently than it does.

I think the former is a reasonable expectation, so I don't believe there's anything to fix here.

**Comment by lolbinarycat** [comment]
you misunderstand my intent.

i know that this is invalid syntax, but usually, in interactive mode, if you hit enter when the input is syntactically invalid, you just get a newline, so you can keep editing.

**Comment by faho** [comment]
You have a syntax error that needs to be fixed *before* the cursor. It can't be fixed by appending more.

We do the same e.g. with `if true; echo )`.

**Comment by lolbinarycat** [comment]
> You have a syntax error that needs to be fixed before the cursor

no, i have a syntax error that needs to be fixed before *end of line*.

if you put the cursor before the syntax error and hit enter, i still get the error.

while it can't be fixed by appending, it can be fixed by insertion, which is what i was trying to do.  unfortunately what i was trying to insert involved a newline, so it didn't let me.

**Comment by zanchey** [comment]
I think the problem is that the error is wrong.

Hitting return after `echo a & |` gives "fish: Expected a string, but found a pipe" (which I guess makes sense - you can only  background a job, not a part of a pipeline).
